### ISSUE TITLE
Is the Engine cross-platform?

### ISSUE THREAD

**Issue by damian-tomczak** [jglrxavpok/Carrot#71]
Hi, I have a question. Is the engine cross-platform?
I ask because I got a lot of weird errors.
![image](https://user-images.githubusercontent.com/43729842/227800938-50c37540-00a6-421b-a290-bd6564a88798.png)

**Comment by jglrxavpok** [comment]
Hello!

The engine works only on Windows currently, but I have never seen these
errors.
How have you setup the project?

Le dim. 26 mars 2023 à 21:54, Damian Tomczak ***@***.***> a
écrit :

**Comment by damian-tomczak** [comment]
@jglrxavpok 

git clone https://github.com/jglrxavpok/Carrot.git --recursive
cd Carrot
mkdir build
cd build
[Installing NsightAftermath and adding the path to the environment variables]
cmake .. . -DCMAKE_PREFIX_PATH=C:\sdks\lua
[Build the whole solution]
Every error and warning: [errors.txt](https://github.com/jglrxavpok/Carrot/files/11073080/errors.txt)

The last errors suggest the problem that the engine was written too much for specific environment. 

btw on what headset are you testing this?

**Comment by jglrxavpok** [comment]
The first build process is a bit complicated, but since I'm the only one working on the engine, it has not been a problem until now.

To test it again, I've tried a fresh clone + build with Visual Studio 16.11.9 (I usually use CLion).

The build process *is* complicated, but CLion handles everything way better than Visual.

1.
You first need to build OpenXR-SDK-Source and OpenXR-Hpp.
See the wiki for the steps: https://github.com/jglrxavpok/Carrot/wiki/Setting-up-a-game-project#if-you-want-to-make-a-vr-game-building-openxr
The wiki is a bit outdated, OpenXR is mandatory now (among other things).
Build "openxr_loader".

2.
Then I recreated the CMake project because I built OpenXR after creating the project...

3.
In openxr_handles_forward, I've had to comment these lines:
![image](https://user-images.githubusercontent.com/5427890/227805346-3bd299e7-939d-4c73-9d35-9298e282128d.png)


4. (not necessary with CLion)
*If building in Debug:*
Then I had to modify the linker input to reference `..\zlib-install\lib\zlibstaticd.lib` instead of `..\zlib-install\lib\zlibstatic.lib` (this is an issue I have not found time to fix yet), therefore using the debug version.
Same goes for `..\..\thirdparty\OpenXR-SDK-Source\build\win64\src\loader\Debug\openxr_loaderd.lib` (debug+d) which has to replace `..\..\thirdparty\OpenXR-SDK-Source\build\win64\src\loader\Release\openxr_loader.lib`

5. (not necessary with CLion)
You need to set the working directory to the build folder for the engine to load properly. `$(SolutionDir)` instead of `$(ProjectDir)`

6. (not necessary with CLion)
Change `Engine.cpp`:
```cpp
Carrot::Engine::SetterHack::SetterHack(Carrot::Engine* e) {
    Carrot::Engine::instance = e;
    Carrot::IO::Resource::vfsToUse = &e->vfs;
    auto exePath = std::filesystem::current_path();//Carrot::IO::getExecutablePath();
    e->vfs.addRoot("engine", exePath/*.parent_path()*/);
    std::filesystem::current_path(exePath/*.parent_path()*/);
}
```
The engine uses the EXE path to add the "engine" mountpoint to the VFS, but Visual does not put it directly inside the build dir.

7.
Reopening the project and trying to build `Peeler` (the editor) worked. You will get MANY windows stacked over one another (no default layout). `BasicSprites` did not compile because I might have forgotten about it when adding precompiled header support.

**Comment by jglrxavpok** [comment]
I'm using the Valve Index to test my VR code, but I have not checked how it behaves recently. But I hope it should still work.

**Comment by damian-tomczak** [comment]
@jglrxavpok 
Please don't be annoyed by stupid questions, but can you expand the fifth point?

> (not necessary with CLion)
You need to set the working directory to the build folder for the engine to load properly. $(SolutionDir) instead of $(ProjectDir)

Because there are a lot of similar projects, and I don't know about which you are talking.

But without even fifth point, I'm getting the ~same errors (especially those weird atomic related).

**Comment by jglrxavpok** [comment]
> can you expand the fifth point?

Any project that you want to launch. I would advise to use **Peeler** (not Peeler-lib or any name similar) to make sure everything works.
Actually, I found a way to tell CMake to generate vsproj with the correct working directory so that should be fixed with commit df4281eef28fee632feaf3afd813b32849feddb7 .

* Missing library?

I tried with your cmake command, and I don't get any the same errors. Though I do not see `openxr_loader.lib` inside the input .lib files for the project "Peeler", do you see it? I use `cmake .. -A x64 -G "Visual Studio 16 2019"` for reference


* Which version of Visual Studio do you use?

It seems 2022 17.5 has changed some things around C11 atomics which *seems* related to your errors?
https://devblogs.microsoft.com/cppblog/c11-atomics-in-visual-studio-2022-version-17-5-preview-2/

I currently use Visual Studio 2019 16.11.9, which may explain why I don't see the same errors.

* stdatomic.h workaround

It looks like the issue is mainly because of OpenEXR (which to be honest gave me more problems than it solves) which finds stdatomic.h and tries to compile with it.
Try to comment these defines inside `thirdparty\openexr\src\lib\OpenEXRCore\chunk.c` and `thirdparty\openexr\src\lib\OpenEXRCore\internal_structs.h`
![image](https://user-images.githubusercontent.com/5427890/228015972-e79ee7c0-1cc0-41fb-81cc-7d1e12639869.png)


* Are you building the entire solution or only the project?

It looks like you are building the entire solution, could you try with a single project? Highly possible I've neglected some projects because I don't use them often.

# Summary of my questions
1. Can you see `openxr_loader.lib` inside your input libraries for **Peeler**?
2. Which version of Visual Studio do you use?
3. Can you try the workaround for stdatomic I gave you?
4. Are you building the entire solution or only a single project?

**Comment by jglrxavpok** [comment]
By the way, I found how I fixed the zlibd issue on my end...

Inside thirdparty/openexr (don't copy yet, read the rest of the message)
```
diff --git a/cmake/OpenEXRSetup.cmake b/cmake/OpenEXRSetup.cmake
index d3e1998c..fe155c6d 100644
--- a/cmake/OpenEXRSetup.cmake
+++ b/cmake/OpenEXRSetup.cmake
@@ -220,6 +220,13 @@ if(OPENEXR_FORCE_INTERNAL_ZLIB OR NOT TARGET ZLIB::ZLIB)
     set(zlibstaticlibname "z")
   endif()

+  if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+    if(WIN32)
+      set(zliblibname "zlibd")
+      set(zlibstaticlibname "zlibstaticd")
+    endif()
+  endif()
+
```
I've updated the commit of OpenEXR I depend on and I think this is no longer needed.

About stdatomic.h
It seems OpenEXR noticed the issue when Visual 2022 was released and has since been fixed, 
https://github.com/AcademySoftwareFoundation/openexr/pull/1237
I've updated which commit my engine depends on, so I hope both issues should be fixed.



TL;DR: try with commit 6003411451224388a3331b20276cba38500f595a

**Comment by damian-tomczak** [comment]
Okey, you have a right with atomic, on 2019 I don't have already errors related to atomics.
In the case of compiling only Peeler I'm getting these errors(and of course after git pull)
[peeler.txt](https://github.com/jglrxavpok/Carrot/files/11083052/peeler.txt)

**Comment by damian-tomczak** [comment]
Okey I have set up the whole project from the base again:
I'm checking now only peeler project
I have to still using VS 2019 because errors from OPENEXR disappeared but appeared errors similar error about chrono.
I'm getting now only three errors:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C1083	Cannot open include file: 'openxr/openxr_platform.h': No such file or directory	Engine-Base	C:\Users\Damian\Desktop\Carrot\engine\engine\vr\includes.h	9	
Error	C1083	Cannot open include file: 'openxr/openxr_platform.h': No such file or directory	Peeler-Lib	C:\Users\Damian\Desktop\Carrot\engine\engine\vr\includes.h	9	
Error	C1083	Cannot open include file: 'openxr/openxr_platform.h': No such file or directory	Peeler	C:\Users\Damian\Desktop\Carrot\engine\engine\vr\includes.h	9	
```

**Comment by damian-tomczak** [comment]
> Okey I have set up the whole project from the base again: I'm checking now only peeler project I have to still using VS 2019 because errors from OPENEXR disappeared but appeared errors similar error about chrono. I'm getting now only three errors:
> 
> ```
> Severity	Code	Description	Project	File	Line	Suppression State
> Error	C1083	Cannot open include file: 'openxr/openxr_platform.h': No such file or directory	Engine-Base	C:\Users\Damian\Desktop\Carrot\engine\engine\vr\includes.h	9	
> Error	C1083	Cannot open include file: 'openxr/openxr_platform.h': No such file or directory	Peeler-Lib	C:\Users\Damian\Desktop\Carrot\engine\engine\vr\includes.h	9	
> Error	C1083	Cannot open include file: 'openxr/openxr_platform.h': No such file or directory	Peeler	C:\Users\Damian\Desktop\Carrot\engine\engine\vr\includes.h	9	
> ```

Okey, it was because of me.
But now I'm getting a new list of errors (even worse...)
[newpeeler.txt](https://github.com/jglrxavpok/Carrot/files/11083279/newpeeler.txt)

**Comment by jglrxavpok** [comment]
Make sure you have compiled OpenXR-SDK-Source before compiling the engine.
https://github.com/jglrxavpok/Carrot/wiki/Setting-up-a-game-project#if-you-want-to-make-a-vr-game-building-openxr
OpenXR is NO LONGER optional contrary to what the article currently says.

**Comment by jglrxavpok** [comment]
Sorry I actually answered the wrong message.
I have no idea for these errors for now.

Le mar. 28 mars 2023 à 22:04, Xavier Niochaut ***@***.***> a
écrit :

> Make sure you have compiled OpenXR-SDK-Source before compiling the engine.
>
> https://github.com/jglrxavpok/Carrot/wiki/Setting-up-a-game-project#if-you-want-to-make-a-vr-game-building-openxr
> OpenXR is NO LONGER optional contrary to what the article currently says.
>

**Comment by jglrxavpok** [comment]
Just tried:

- Update to Visual Studio 2022 17.5.3
- git clone
- build OpenXR-SDK-Source
- build OpenXR-Hpp
- `cmake -A x64 .. -G "Visual Studio 17 2022" in carrot build directory
- point to `..\..\thirdparty\OpenXR-SDK-Source\build\win64\src\loader\Debug\openxr_loaderd.lib` in Peeler library (instead of Release version)
- fix errors in openxr_handles_forward.hpp
- comment std::hash specialization for std::filesystem::path
- change static_assert to verify inside Action.hpp, line 208
- change Engine.cpp with the current_path fix I gave a few messages ago
- launch Peeler


Though since I updated to VS 2022, CLion gave me the same errors on my original version of the engine, I've had to rebuild.
**I'm guessing precompiled headers might create the issues you currently have.**

**Comment by damian-tomczak** [comment]
Okay, I repeated the whole process and still I'm getting errors. In about a few days(I'm starving for the lack of the time), I will record the whole process - maybe I'm doing something stupid.
